<a href="https://colab.research.google.com/github/lilasu086/Individual_CodingProject/blob/main/SupplyChain_PricingAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Part I: Pricing Analysis - Optimal pricing for weekend vs. weekday

In [ ]:
import pandas as pd
import numpy as np

demand = pd.read_excel('/content/demand_data_full.xlsx')
demand

price  demand  weekend
0     10.5      14        0
1     15.0      12        0
2      6.0      16        0
3     12.0      32        1
4     21.0       8        0
..     ...     ...      ...
255   14.4      12        0
256   21.0      18        1
257   10.5      14        0
258    6.0      16        0
259   12.0      14        0

[260 rows x 3 columns]

In [ ]:
demand.isna().sum()

price      0
demand     0
weekend    0
dtype: int64

### 1.a

In [ ]:
!pip install stargazer

In [ ]:
import statsmodels.api as sm
from stargazer.stargazer import Stargazer

ols_1 = sm.OLS.from_formula("demand ~ price + weekend", data=demand).fit()
Stargazer([ols_1])

[Answer 1.a]

Based on the OLS summary report, it is evident that the regression model, which includes the variables `price` and `weekend`, accounts for 94% of the variance in the `demand` variable. Additionally, the analysis reveals that for every \$1 decrease in price, demands are expected to increase by 0.785. Furthermore, during weekend, demands are projected to increase by 16.019. It is important to note that all coefficients in the regression are statistically significant(they have three stars in the regression summary).

### 1.b

In [ ]:
ols_1.params

Intercept    22.733062
price        -0.785299
weekend      16.019304
dtype: float64

In [ ]:
pi_1 = lambda x: -(ols_1.params[0] + ols_1.params[1]*x + ols_1.params[2]*0) * (x-4)

In [ ]:
from scipy.optimize import minimize_scalar
optimal_price_1 = minimize_scalar(pi_1, method='brent')
optimal_price_1.x

16.474137156593187

In [ ]:
optimal_price_1

 message: 
          Optimization terminated successfully;
          The returned value satisfies the termination criteria
          (using xtol = 1.48e-08 )
 success: True
     fun: -122.19580005551767
       x: 16.474137156593187
     nit: 4
    nfev: 9

In [ ]:
demand.price.min()

4.5

In [ ]:
demand.price.max()

22.5

In [ ]:
print(f'(Min) {demand.price.min()}', ' < ', round(optimal_price_1.x,2), ' < ', f'{demand.price.max()} (Max)')
demand.price.min() < optimal_price_1.x < demand.price.max()

(Min) 4.5  <  16.47  <  22.5 (Max)


True

[Answer 1.b]

The optimal weekday price (when w = 0) is \$16.4741 and its corresponding profit is \$122.1958.

This optimal weekday price falls within the credible range, meaning it lies between the maximum and minimum price.

### 1.c

In [ ]:
pi_2 = lambda x: -(ols_1.params[0] + ols_1.params[1]*x + ols_1.params[2]*1) * (x-4)

In [ ]:
from scipy.optimize import minimize_scalar
optimal_price_2 = minimize_scalar(pi_2, method='brent')
optimal_price_2.x

26.673625915080788

In [ ]:
optimal_price_2

 message: 
          Optimization terminated successfully;
          The returned value satisfies the termination criteria
          (using xtol = 1.48e-08 )
 success: True
     fun: -403.7171544152613
       x: 26.673625915080788
     nit: 4
    nfev: 9

In [ ]:
print(f'(Min) {demand.price.min()}', ' < ', round(optimal_price_2.x,2), ' < ', f'{demand.price.max()} (Max)')
demand.price.min() < optimal_price_2.x < demand.price.max()

(Min) 4.5  <  26.67  <  22.5 (Max)


False

[Answer 1.c]

The optimal weekend price (when w = 1) is \$26.6736 and its corresponding profit is \$403.7172. The optimal weekend price is not within the credible range.

This optimal weekday price does not fall within the credible range, meaning it does not lie between the maximum and minimum price.

## Part II: Pricing Analysis - New vs Refurbished Product

In [ ]:
refurb = pd.read_csv('/content/refurb_data.csv')
refurb

dates  new_price  refurb_price  choice
0    2014-06-01        400         264.0  refurb
1    2014-06-01        400         264.0  refurb
2    2014-06-01        400         264.0  refurb
3    2014-06-01        400         264.0     new
4    2014-06-01        400         264.0     new
..          ...        ...           ...     ...
495  2014-06-30        300         222.0     new
496  2014-06-30        300         222.0     new
497  2014-06-30        300         222.0     new
498  2014-06-30        300         222.0     new
499  2014-06-30        300         222.0  refurb

[500 rows x 4 columns]

In [ ]:
refurb['dates'] = pd.to_datetime(refurb['dates'])
refurb['new_price'] = refurb['new_price'].astype('float')
refurb['refurb_price'] = refurb['refurb_price'].astype('float')
refurb.info()
refurb.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   dates         500 non-null    datetime64[ns]
 1   new_price     500 non-null    float64       
 2   refurb_price  500 non-null    float64       
 3   choice        500 non-null    object        
dtypes: datetime64[ns](1), float64(2), object(1)
memory usage: 15.8+ KB


dates  new_price  refurb_price  choice
0 2014-06-01      400.0         264.0  refurb
1 2014-06-01      400.0         264.0  refurb
2 2014-06-01      400.0         264.0  refurb
3 2014-06-01      400.0         264.0     new
4 2014-06-01      400.0         264.0     new
5 2014-06-01      400.0         264.0  refurb
6 2014-06-01      400.0         264.0  refurb
7 2014-06-01      400.0         264.0     new
8 2014-06-01      400.0         264.0     new
9 2014-06-01      400.0         264.0  refurb

In [ ]:
# change refurb into 1 and new into 0
refurb['refurb_sales'] = refurb['choice'].replace({'refurb': 1, 'new': 0})

In [ ]:
refurb_new = refurb.groupby('dates').agg(refurb_price=('refurb_price', 'first'),
                                    new_price=('new_price', 'first'),
                                    refurb_sales=('refurb_sales', 'sum')).reset_index()
refurb_new

dates  refurb_price  new_price  refurb_sales
0  2014-06-01         264.0      400.0             9
1  2014-06-02         305.0      400.0             3
2  2014-06-03         306.0      400.0             6
3  2014-06-04         283.0      400.0             7
4  2014-06-05         305.0      400.0             8
5  2014-06-06         278.0      400.0             7
6  2014-06-07         337.0      400.0             3
7  2014-06-08         286.0      400.0             8
8  2014-06-09         335.0      400.0             5
9  2014-06-10         277.0      400.0             7
10 2014-06-11         277.0      350.0             8
11 2014-06-12         273.0      350.0             6
12 2014-06-13         204.0      350.0             6
13 2014-06-14         298.0      350.0             1
14 2014-06-15         242.0      350.0             4
15 2014-06-16         228.0      350.0             7
16 2014-06-17         293.0      350.0             4
17 2014-06-18         267.0      350.0             6
18 2014-06-19         299.0      350.0             3
19 2014-06-20         223.0      350.0             6
20 2014-06-21         161.0      300.0             9
21 2014-06-22         217.0      300.0             6
22 2014-06-23         175.0      300.0             2
23 2014-06-24         163.0      300.0            10
24 2014-06-25         238.0      300.0             6
25 2014-06-26         242.0      300.0             1
26 2014-06-27         151.0      300.0             7
27 2014-06-28         215.0      300.0             6
28 2014-06-29         220.0      300.0             3
29 2014-06-30         222.0      300.0             6

In [ ]:
refurb_fit_1 = sm.OLS.from_formula(formula='refurb_sales ~ refurb_price + new_price', data=refurb_new).fit()
Stargazer([refurb_fit_1])

In [ ]:
refurb_fit_1.params

Intercept      -0.701876
refurb_price   -0.044835
new_price       0.050579
dtype: float64

In [ ]:
pi_3 = lambda x: -(refurb_fit_1.params[0] + refurb_fit_1.params[1]*x + refurb_fit_1.params[2]*300) * (x-150)

In [ ]:
from scipy.optimize import minimize_scalar
optimal_price_3 = minimize_scalar(pi_3, method='brent')
optimal_price_3

 message: 
          Optimization terminated successfully;
          The returned value satisfies the termination criteria
          (using xtol = 1.48e-08 )
 success: True
     fun: -334.6266140552377
       x: 236.3920313609819
     nit: 4
    nfev: 10

In [ ]:
pi_4 = lambda x: -(refurb_fit_1.params[0] + refurb_fit_1.params[1]*x + refurb_fit_1.params[2]*350) * (x-150)

In [ ]:
from scipy.optimize import minimize_scalar
optimal_price_4 = minimize_scalar(pi_4, method='brent')
optimal_price_4

 message: 
          Optimization terminated successfully;
          The returned value satisfies the termination criteria
          (using xtol = 1.48e-08 )
 success: True
     fun: -588.7710949351213
       x: 264.5952685352089
     nit: 4
    nfev: 10

In [ ]:
pi_5 = lambda x: -(refurb_fit_1.params[0] + refurb_fit_1.params[1]*x + refurb_fit_1.params[2]*400) * (x-150)

In [ ]:
from scipy.optimize import minimize_scalar
optimal_price_5 = minimize_scalar(pi_5, method='brent')
optimal_price_5

 message: 
          Optimization terminated successfully;
          The returned value satisfies the termination criteria
          (using xtol = 1.48e-08 )
 success: True
     fun: -914.240451311913
       x: 292.7985057094407
     nit: 4
    nfev: 10

[Answer 2]

When the new price is \$300, the optimal refurbished price is \$236.3920 and the profit is \$334.6266.

When the new price is \$350, the optimal refurbished price is \$264.5953 and the profit is \$588.7711.

When the new price is \$400, the optimal refurbished price is \$292.7985 and the profit is \$914.2405.

Based on the findings above, it's evident that with an increase in the new price, there's a corresponding rise in the refurbished price, leading to higher profits. This trend can be attributed to the perceived higher value of the product when the new price increases. Consequently, the refurbished price follows suit, reflecting the increased value. As both prices experience an upward trend, it naturally results in higher profitability.



## Part III: Pricing Analysis - Two-Segment pricing

### With no segmentation

In [ ]:
pi_6 = lambda x: -(10000-800*x) * (x-5)

In [ ]:
from scipy.optimize import minimize_scalar
optimal_price_6 = minimize_scalar(pi_6, method='brent')
optimal_price_6

 message: 
          Optimization terminated successfully;
          The returned value satisfies the termination criteria
          (using xtol = 1.48e-08 )
 success: True
     fun: -11249.999999999998
       x: 8.750000000000005
     nit: 4
    nfev: 9

In [ ]:
print('the optimal price in the single market is: ' + '$' + str(round(optimal_price_6.x,3)))
print('the optimal profit in the single market is: ' + '$' + str(-round((optimal_price_6.fun),1)))

the optimal price in the single market is: $8.75
the optimal profit in the single market is: $11250.0


In [ ]:
# profit with no segmentation
print('the optimal profit in the non-segmentation market is: ' + '$' + str(-round((optimal_price_6.fun),1)))

the optimal profit in the non-segmentation market is: $11250.0


### With segmentation

In [ ]:
segment_price = 7

In [ ]:
# Low WTP segment size
low_wtp_size = round((segment_price/12.5)*10000,1)
low_wtp_size

5600.0

In [ ]:
pi_7 = lambda x: -(5600-800*x) * (x-5) if x <= 7 else -0 * (x-5)

In [ ]:
from scipy.optimize import minimize_scalar
optimal_price_7 = minimize_scalar(pi_7, method='brent')
optimal_price_7

 message: 
          Optimization terminated successfully;
          The returned value satisfies the termination criteria
          (using xtol = 1.48e-08 )
 success: True
     fun: -799.9999999999998
       x: 6.000000000000001
     nit: 6
    nfev: 11

In [ ]:
print('the optimal price in the low demand market is: ' + '$' + str(round(optimal_price_7.x,3)))
print('the optimal profit in the low demand market is: ' + '$' + str(-round((optimal_price_7.fun),1)))

the optimal price in the low demand market is: $6.0
the optimal profit in the low demand market is: $800.0


In [ ]:
# High WTP segment size
high_wtp_size = round((12.5-segment_price)/12.5 *10000,1)
high_wtp_size

4400.0

In [ ]:
pi_8 = lambda x: -4400*(x-5) if x <= 7 else -(10000-800*x) * (x-5)

In [ ]:
from scipy.optimize import minimize_scalar
optimal_price_8 = minimize_scalar(pi_8, method='brent')
optimal_price_8

 message: 
          Optimization terminated successfully;
          The returned value satisfies the termination criteria
          (using xtol = 1.48e-08 )
 success: True
     fun: -11250.0
       x: 8.74999999999987
     nit: 8
    nfev: 13

In [ ]:
print('the optimal price in the high demand market is: ' + '$' + str(round(optimal_price_8.x,3)))
print('the optimal profit in the high demand market is: ' + '$' + str(-round((optimal_price_8.fun),1)))

the optimal price in the high demand market is: $8.75
the optimal profit in the high demand market is: $11250.0


In [ ]:
# profit with segmentation
print('the optimal profit in the segmentation market is: ' + '$' + str(-round((optimal_price_8.fun),1) + -round((optimal_price_7.fun),1)))

the optimal profit in the segmentation market is: $12050.0


[Answer 3]

Without segmentation, profits amount to \$11,250, whereas with segmentation, profits increase to \$12,050. This outcome suggests that segmentation leads to higher profits. The rationale behind this improvement lies in the fact that different market segments exhibit varying average Willingness To Pay (WTP). Consequently, companies can optimize their profits by tailoring their pricing strategies to match the distinct WTP levels of each segment.

Notably, the price and demand function remain unchanged in the high WTP segmentation, resembling those in the single-priced market. However, the enhanced profits in the segmented market predominantly stem from the low WTP segmentation, indicating the effectiveness of segmentation in maximizing profitability.